ARD-4020 Implemented state stream tooling with tests (WIP)

anton164 opened the pull request:
The pattern for creating state streams is one of the few missing pieces in ardoq-front to fully make use of reducers from rxbeach, so I thought I would give this a go.

The goal is for this to replace `createReducedStateStream` in ardoq-front (and hopefully most cases of `createStateStream` now that we have data reducers).

### The proposal

  **reduceState guarantees that:**
  - a value is emitted upon subscription with startWith (state streams can't be lazy)
  - the stream producer is shared, if there are multiple subscriptions
    the reducers will only run once
  - the state stream producer lives on "forever" throughout the app lifecycle,
    subsequent subscriptions will always get the previous state calculation
    even though there are no subscribers (i.e. the stream is not refcounted)

  **WIP:**
  - a state stream forwards errors to an error subject and continues emitting values after an error occurs

_All of the above "properties" have unit tests in **reduceState.spec.ts**_

### Example

With this tooling you can define a state stream like so:
```typescript
const defaultState = 1;
const handleOne = reducer(incrementOne, incrementOneHandler);
const handleMany = reducer(
  incrementMany,
  (accumulator: number, increment) => accumulator + increment
);
const state$ = action$.pipe(
  reduceState('testStream', defaultState, [
    handleOne,
    handleMany,
  ])
);
```

anton164 commented on src/stateStream.ts:
```diff
@@ -0,0 +1,82 @@
+ * @see rxbeach.combineReducers
+ * @returns An stream that emits the reduced state
+ */
+export const stateStream = <State>(
```
Not 100% settled on this signature. Since there are 4 args, perhaps it's better to take an object? In the end I did not use an object to stay consistent with the rest of rxbeach and since none of the arguments are optinoal

totokaka replied:
I believe a better signature would be `<State>(name: string, defaultState: State, reducers: RegisteredReducer<State, any>[]) => OperatorFunction<UnknownAction, State>`. This of course means this function is a stream operator.

Having this function be a stream operator is more composable, and will make it much nicer to use it together with `useNamespace`, for instance. It also avoids the problem of too many arguments.

anton164 replied:
Gave this a bit more thought and I agree :) 

My initial "fear" was that making it composable would allow someone to misuse it, but I think we should aim to avoid misuse in other ways. Furthermore, your example of `useNamepace` clearly demonstrates why making it composable as an operator function is important, and I can imagine there might be other uses too.

anton164 commented on src/stateStream.ts:
```diff
@@ -0,0 +1,82 @@
+    tag(name)
+  );
+
+export const stateStreamWithErrorHandling = <State>(
```
This is obviously WIP, as I would like for this error handling to be a part of `stateStream`, however it turned out to be quite tricky to implement the desired behaviour (see the last test).

totokaka replied:
Yeah, this implementation is horrendous :fearful: 

I'll give it a spin and see if I can do something better. 

I hope the plan is to merge this and the `stateStream` function?

anton164 replied:
> I hope the plan is to merge this and the stateStream function?
Yes :) I split it out because it became so ugly 😅 Error handling and continuing, while emitting the correct things is a bit tricky with observables.

anton164 commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  })
+);
+
+test('stateStream subscribers should only see one emission if reducers emit synchronously', async t => {
```
This is a "good" side-effect, but per @chriskr 's suggestion, we should understand exactly why it happens 😅

totokaka replied:
I think this test is misguided. `toPromise` resolves once the stream completes. In this test `state$` will complete when `action$` completes.
I would guess `state$` actually emits the `defaultState`, but it's not caught because of the behavior of `toPromise`.

Anyways, let's not use `toPromise` in the tests :)

anton164 replied:
True, it's not just misguided, it's completely at odds with the earlier test `reduceState should call reducers synchronously if they emit at once` 🤦‍♂ 

Removed now

chriskr commented on src/stateStream.ts:
```diff
@@ -0,0 +1,82 @@
+ * The state stream will always emit its last value upon subscription.
+ * On the first subscription to the state stream, it will emit `defaultState`.
+ *
+ * The state stream is not ref counted, which means that the underlying
```
That means the stream never stops, continues to calculate state no matter if the state is used or not? Is this meant to be used for view state streams? Seems the wrong pattern to me? A view state stream should always calculate the valid state at the moment of connect?

totokaka replied:
This is a very good point, and warrants a proper discussion. I agree that the "view state stream" should be destroyed once the view is unmounted. If something should persist when the view is unmounted, it should go in another stream.

anton164 replied:
For sure a good point! I think we should have a follow-up discussion on Wednesday.

The reasoning behind this decision was that I would like to try to avoid introducing an [isTemporary](https://github.com/ardoq/ardoq-front/blob/ab3dc69071c59595669173df6ac975d565801497/src/js/ardoq/streams/utils/stateStreamUtils.ts#L31) flag akin to what we have in `createStateStream` in ardoq-front.

From my observations in the codebase (and based on past experience with Redux), I think it's much easier to reason about state streams that aren't bound to a view's lifecycle, but rather the application's lifecycle. 

Three somewhat orthogonal reasons for this:
- **persisting state beyond a view's lifecycle:** a view might at some point unmount unexpectedly, but that doesn't necessarily mean that we wish to reset the state. In many cases, the best user experience is when the view state persist beyond the view's lifecycle. For example when you are editing a gremlin query, but click to open the workspaces module, then go back to editing the gremlin query.
- **reducing state for the view when the view is not mounted:** in some cases, a state stream should be ready to reduce actions before the view is mounted. I.e. we might wish to switch to a view within a different module and set some state. If we need to wait until the view is mounted to set the state, it can lead to some nasty code.
- **consistency and simplicity:** if we can avoid having two types of reduced state streams, one that resets and another that persists throughout the application lifecycle, I think that will simplify application development.

The first problem can be solved using **refCount: false**, while the second problem needs some kind of [global registry that sets up "global"/app-wide state streams](https://github.com/ardoq/ardoq-front/blob/dfe5c7312bd041858fb85a59e800a19647f7eccf/src/js/ardoq/startStateStreams.ts), similar to what is necessary for routines.

As both of you point out, there is also a third use-case: creating reduced state that is bound to a view, which only sets up once the view is mounted and completes once the view unmounts. This is especially true for our visualizations, since we don't want them to spend time building their view models when they are inactive.

However, here we must be careful to not mix two things - **reduced state** and **view model calculation**. 

IMO, view model calculation should obviously be bound to when a view mounts, since it is only necessary for the view when it's active, while **reduced state** can and should be bound to the application's lifecycle.

It's _highly possible_ that there's some important points that I'm missing, which requires us to have reducers that are bound to a view. That's why I think we should discuss this further :)

totokaka replied:
Here's a problem case:
> The view model streams of visualizations should be reduced streams, have the same life-cycle as their view (React) components and be namespaced to each mounted instance. 

In this case the only thing that is in the WIP implementation of `reduceState` that we don't need / don't want is `shareReplay`.

I think one take-away from this is that we don't really have a use case where we need `shareReplay` without `refCount: false`. So maybe we could define a very simple alias for `shareReplay({ refCount: false, bufferSize: 1 })` named something like `persistAndReplay`? Then reduced state streams would need two operators, but we're free to reuse `reduceState` when creating the tooling for view model streams.

anton164 replied:
True, there are some obvious examples from our current views where we wish to have the reset pattern, but still keep reducers around 🤔 I was trying hard to entertain the idea that we can live without them, but that's a bit detached from reality 😄 

However, I don't think it's a good idea to add a `persistAndReplay` operator, since it would most require state stream definitions to be:
```typescript
const state$ = action$.pipe(
   reduceState(),
   persistAndReplay()
);
// instead of simply
const state$ = action$.pipe(
   reduceState()
);
```

IMO, we should either add some option to `reduceState` that allows to configure this behaviour, or create a separate interface, i.e. `reduceViewState`. If we roll with `reduceViewState` it might make sense to also rename `reduceState`.

anton164 replied:
Added another test, which shows an interesting consequence of `({ refCount: false })`:
https://github.com/ardoq/rxbeach/blob/231ad441add1b662c430a893affbe516b8b10e71/src/reduceState.spec.ts#L111-L127

Perhaps, the best solution is actually to roll with `refCount: true` instead. This allows for more fine-grained control as we can distinguish between "permanent" and "temporary" streams by having a registry. 

If you need a "local" stream that resets, you simply avoid registering it in `startGlobalStateStreams` registry.

If you need a "local" stream per view instance, you have a simple method:
```typescript
const createInstanceStream (action$, instanceId) => action$.pipe(
   withNamespace(instanceId),
   reduceState(...)
);
```

anton164 replied:
Finally, I have pushed a commit that switches the implementation to `refCount: true`. I think that's the best way to go :) I've requested a new review

totokaka commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  'stateStream should emit initial state synchronously upon subscription',
+  marbles(m => {
+    const defaultState = 3;
+    const action$ = m.hot('');
```
```suggestion
    const action$ = m.hot('  -');
```
To explicitly indicate it doesn't emit anything in the time frame the other stream emits `3`

totokaka commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+);
+
+test(
+  'stateStream should call reducers synchronously if they emit at once',
```
I don't think this test belongs here - as it's just testing the functionality of `combineReducers`

anton164 replied:
Hmm, not sure if I agree because it is testing the combination of `startWith` and `combineReducers`. 

I renamed the test to: `reduceState should emit defaultState first and call reducers synchronously if they emit at once`

totokaka commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  t.is(await state$.toPromise(), 2);
+});
+
+test('stateStream should call reducer once when there are multiple subs', async t => {
```
```suggestion
test('stateStream should call reducer once when there are multiple subs', t => {
```
An alternative way to test this would be to do it like the test `stateStream subscribes action$`, but check that `action$` still only have one subscription when the `state$` has two.

anton164 replied:
True, but I think this way is more explicit. I could add the test that ensures `action$` is only subscribed once, but that might be superfluous?

totokaka commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  sub2.unsubscribe();
+});
+
+test('stateStream should never reset to its default state', async t => {
```
I'm a bit unsure about how this test is implemented. We should look into if and how it is possible to test this behavior with marble tests.

anton164 replied:
I gave up the first time, but finally found out how to do it :)

The key was to pass a subscriber as a second argument to `m.expect`

Relevant: [RTFM 😄  ](https://ncjamieson.com/marble-testing-rtfm/)

totokaka commented on src/stateStream.spec.ts:
```diff
@@ -0,0 +1,134 @@
+    const handleWord = reducer(word$, () => {
+      throw 'error';
+      // eslint-disable-next-line no-unreachable
+      return 1;
```
```suggestion
    const handleWord = reducer(word$, (): number => {
      throw 'error';
```
I think this would work?

And this test would also be a bit more concise if we didn't use stream reducers (`handleWord`), but instead followed the style of this test: https://github.com/ardoq/rxbeach/blob/master/src/reducer.spec.ts#L95

anton164 replied:
good suggestion, thanks 👍

anton164 commented:
@davidnegley / @chriskr  / @totokaka  thanks for the reviews!

@totokaka's discussion points:
> The name - I believe we shouldn't reserve the term "state stream" for reduced streams. Is "reduced state stream" a better name? At least it means we can avoid conflicts if we would like to add concepts with names like "derived state stream" later.
> 
> The signature - I strongly believe having this as an operator function is the way to go. See my arguments in the comments.

I agree to both points. I've renamed it to `reduceState` and refactored it to be an operator function.

> Two smaller thing:
>
> It seems the concept of incrementing has caught on as a nice test reducer. Maybe we should move these reducers, their handler functions and the actions, words and numbers object to their own file? It should probably be a separate PR or at least a different commit.

Agree, I can do a folllow-up once this is merged

> If the commit message starts with feat(state stream): we can get nice change logs :) I'm also unsure if it gives any value to have our internal issue number in the commit message.

Agree 👍 

I've addressed the rest of your feedback in the threads and/or via fixup commits.

totokaka commented on src/reduceState.spec.ts:
```diff
@@ -0,0 +1,128 @@
+
+    m.expect(state$).toBeObservable(expected$);
+    m.expect(state$, sub1).toBeObservable(sub1Expected$);
+    m.expect(state$, sub2).toBeObservable(sub2Expected$);
```
This really nice!
